Add a 'requireForBlockBody' modifier to the 'arrow-parens' rule (fixes #6557)

nfroidure opened the pull request:
The when-brace option allows to require parens for arrow functions whose body is
 wrapped into curly braces like the following:
 ```js
 (a) => {}
 ```

Otherwise, its behavior remains identical then with the as-needed option.

SebastienElet commented:
👍

kaicataldo commented:
Issue is now accepted

kaicataldo commented:
@nfroidure Thanks for contributing to ESLint! We're ready to start moving on this now that the enhancement has been discussed and accepted. The team ended up deciding on a different configuration option than the one you have here - would you be willing to update the PR so that it reflects the issue? Let me know if you have any questions!

nfroidure commented:
@kaicataldo thought you were on it, i can do it as well, just let me know to avoid doing it twice.

kaicataldo commented:
Ah, understand the confusion. As a champion for this enhancement it's up to me to either implement or to help someone else implement it. We'd love your contribution, if you're willing :)

If you don't have the time or are not interested, I can definitely take this on.

nfroidure commented:
@kaicataldo done ;). I amended the commit and renamed it, just tell me if i also have to change the branch name.

nfroidure commented:
@eslintbot For sure, wait a minute ⌚

kaicataldo commented:
No worries on the branch name :) The team will start reviewing and we'll leave comments here, should we feel anything needs to be changed/updated. Haven't had a chance to look at the PR yet but will try to do so ASAP. Thanks for contributing!

btmills commented on docs/rules/arrow-parens.md:
```diff
@@ -49,13 +49,16 @@ a => {}
+
+ This rule has an object option for exceptions to the `"as-needed"` option:
 
 You can set the option in configuration like this:
```
Looks like we don't need this line anymore.

btmills commented on docs/rules/arrow-parens.md:
```diff
@@ -176,4 +179,45 @@ a.then(foo => { if (true) {}; });
+
+### requireForBlockBody
+
+When the rule is set to `"as-needed", { "requireForBlockBody": true }` the following patterns are considered problems:
```
To help our documentation generator, can you change this line to read

> Examples of **incorrect** code for the `{ "requireForBlockBody": true }` option:

The line introducing the correct examples below should be similar, saying "correct" instead of "incorrect".

btmills commented on docs/rules/arrow-parens.md:
```diff
@@ -176,4 +179,45 @@ a.then(foo => { if (true) {}; });
+() => {};
+a => a;
+a.then((foo) => {});
+a.then(foo => { if (true) {}; });
```
Doesn't this example need parens?

btmills commented on lib/rules/arrow-parens.js:
```diff
@@ -20,7 +20,16 @@ module.exports = {
         schema: [
             {
-                enum: ["always", "as-needed"]
+                enum: ["always", "as-needed", "when-brace"]
```
The `"when-brace"` option isn't needed anymore.

btmills commented:
Just a handful of minor tweaks, otherwise LGTM. Thanks @nfroidure!

ljharb commented:
@nfroidure Thank you!! I'm looking forward to enabling this rule in Airbnb's linter config :-)

nfroidure commented:
@btmills thanks for the review, just updated the PR.

btmills commented on docs/rules/arrow-parens.md:
```diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
+
+### requireForBlockBody
+
+Examples of incorrect code for the { "requireForBlockBody": true } option:
```
Oops, it looks like this line and the corresponding "correct" line lost some formatting when you copied them from my previous comment. The words "incorrect" and "correct" need double asterisks, and the code needs backticks:

```md
Examples of **incorrect** code for the `{ "requireForBlockBody": true }` option:
```

nfroidure replied:
@btmills Just fixed it. Sorry.

kaicataldo commented on docs/rules/arrow-parens.md:
```diff
@@ -49,13 +49,14 @@ a => {}
 ## Options
 
-The rule takes one option, a string, which could be either `"always"` or `"as-needed"`. The default is `"always"`.
+This rule has a string option:
```
My preference would be to note at the top level that this also has a second object option and then go into more detail into each option. That way the reader can know at a glance that they should keep going down the page if they're interested in all the options.

kaicataldo commented on docs/rules/arrow-parens.md:
```diff
@@ -49,13 +49,14 @@ a => {}
 
-You can set the option in configuration like this:
+* `"always"` (default) requires parens in all cases.
+* `"as-needed"` allows ommitting parens in some cases.
```
Can we describe here, even at a high level, what these cases are?

kaicataldo commented on docs/rules/arrow-parens.md:
```diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
+/*eslint arrow-parens: [2, "as-needed", { "requireForBlockBody": true }]*/
+/*eslint-env es6*/
+
+(a) => {};
```
Thoughts on adding `a => ({});` to the correct examples?

kaicataldo commented on docs/rules/arrow-parens.md:
````diff
@@ -176,4 +177,44 @@ a.then(foo => { if (true) {}; });
+```
+
+The `"as-needed", { "requireForBlockBody": true }` rule is directly inspired by the Airbnb
+ [JS Style Guide](https://github.com/airbnb/javascript#arrows--one-arg-parens).
````
Thoughts on adding this link in a section at the bottom like many of other rules do? Example: https://github.com/eslint/eslint/blob/master/docs/rules/no-debugger.md

kaicataldo commented on lib/rules/arrow-parens.js:
```diff
@@ -29,18 +38,63 @@ module.exports = {
+            if (
+                requireForBlockBody &&
+                node.params.length === 1 && node.params[0].type === "Identifier" &&
+                bodyToken.type !== "Punctuator" && bodyToken.value !== "{"
```
I personally think `node.body.type !== 'BlockStatement'` is clearer than `bodyToken.type !== "Punctuator" && bodyToken.value !== "{"`. Thoughts?

kaicataldo commented on lib/rules/arrow-parens.js:
```diff
@@ -29,18 +38,63 @@ module.exports = {
+
+            if (
+                requireForBlockBody &&
+                bodyToken.type === "Punctuator" && bodyToken.value === "{"
```
Same comment as above - thoughts on checking if `node.body.type` is 'BlockStatement'? Seems like a good idea to rely on info the parser is giving us, since we can.

kaicataldo commented on lib/rules/arrow-parens.js:
```diff
@@ -29,18 +38,63 @@ module.exports = {
+                requireForBlockBody &&
+                bodyToken.type === "Punctuator" && bodyToken.value === "{"
+            ) {
+                if (token.type !== "Punctuator" || token.value !== "(") {
```
Is this just to short-circuit the check if it's not a punctuator? That's cool! I would have written it `token.type !== "Punctuator" || token.value !== "("`, but I like this :)

kaicataldo commented on tests/lib/rules/arrow-parens.js:
```diff
@@ -18,30 +18,53 @@ var rule = require("../../../lib/rules/arrow-parens"),
+    { code: "(a, b) => {}", options: ["as-needed"], parserOptions: { ecmaVersion: 6 } },
+
+    // "as-needed", { "requireForBlockBody": true }
+    { code: "() => {}", options: ["as-needed", {requireForBlockBody: true}], parserOptions: { ecmaVersion: 6 } },
```
Can we add tests that ensure `(a) => {};` and `a => ({});` are valid?

kaicataldo commented on tests/lib/rules/arrow-parens.js:
```diff
@@ -133,6 +156,44 @@ var invalid = [
+        }]
+    },
+    {
+        code: "(b) => b",
```
This looks like a duplicate of the previous test - am I missing something?

kaicataldo commented on tests/lib/rules/arrow-parens.js:
```diff
@@ -133,6 +156,44 @@ var invalid = [
+            message: requireForBlockBodyMessage,
+            type: type
+        }]
     }
```
Could we add a test that this is invalid: `(a) => ({});`. Don't mean to harp on that one case, but it seems like one that is more likely to break if someone ever touches this in the future. Would also love to see a test for destructured params in the invalid tests!

kaicataldo commented:
Left some suggestions. Thanks so much for working on this @nfroidure!

kaicataldo commented:
@nfroidure Just wanted to check in and see if there's anything I can do to help get this landed. Looks like we need to rebase and resolve a merge conflict as well.

nfroidure commented:
@kaicataldo everything is well just had no time to dive back in this PR. I'll try to give it a look this week.

nfroidure commented:
@kaicataldo just adapted the PR to your recommendations and rebased:
- about the duplicate it already was on the previous tests and was not sure of its usefulness. I removed both.
- i added a test for the `() => ({})` case.
- was not aware of the BlockStatement node type and its implications, just replaced it and it actually fixed the edge case you pointed out, nice catch ;).
- both {a, b} => a + b and [a, b] => a + b leads to a syntax error so i did not test them. Just added some tests to check ({a, b}) => a + b and ([a, b]) => a + b as valid.

ljharb commented:
@nfroidure did you force push it up? the branch check still indicates conflicts

nfroidure commented:
@ljharb i pushed it before rebasing to give you the opportunity to review the changes before the rebase but it appeared to be a minor conflict ;)

kaicataldo commented on lib/rules/arrow-parens.js:
```diff
@@ -29,9 +38,13 @@ module.exports = {
 
         let sourceCode = context.getSourceCode();
 
+
```
Extra newline

kaicataldo commented:
Looks like a rebase is needed, but otherwise this is looking great to me.

nzakas commented:
@kaicataldo feel free to merge if you think this is ready.

kaicataldo commented:
LGTM. Thanks for contributing, @nfroidure!

nfroidure commented:
Was a great pleasure to help on a product i use on a daily basis. Keep up the good work guys 😘
